### PR TITLE
Rook-Ceph CSV templates

### PR DESCRIPTION
**Pull request by davidvossel**
**Description of your changes:**

The ocs-operator would like to consume a CSV template from the rook-ceph operator's container image. This template is used to merge the rook-ceph operator into the ocs-operators unified CSV file.

In addition to the CSV templates, I've also included the rook related CRDs in the container image for extraction. 

**Usage Example** extracting CSV template for OCP from the rook-ceph container image. 

```
ROOK_CEPH_IMAGE=<some-image>
docker run --rm -it --entrypoint=cat $ROOK_CEPH_IMAGE /etc/ceph-csv-templates/rook-ceph-ocp.vVERSION.clusterserviceversion.yaml.in
```

**Usage Example** extract a crd
```
ROOK_CEPH_IMAGE=<some-image>
docker run --rm -it --entrypoint=cat $ROOK_CEPH_IMAGE /etc/ceph-csv-templates/crds/ceph_crd.yaml
```

**Usage Example** List all CRDs. 
```
ROOK_CEPH_IMAGE=<some-image>
docker run --rm -it --entrypoint=ls $ROOK_CEPH_IMAGE -l /etc/ceph-csv-templates/crds
total 16
-rw-rw-r--. 1 root root 2233 Aug 15 21:49 ceph_crd.yaml
-rw-rw-r--. 1 root root  364 Aug 15 21:49 rookcephblockpools.crd.yaml
-rw-rw-r--. 1 root root  378 Aug 15 21:49 rookcephobjectstores.crd.yaml
-rw-rw-r--. 1 root root  410 Aug 15 21:49 rookcephobjectstoreusers.crd.yaml

```

**Checklist:**

- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/master/development-flow.html#submitting-a-pull-request)
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.
- [ ] Pending release notes updated with breaking and/or notable changes, if necessary.
- [ ] Upgrade from previous release is tested and upgrade user guide is updated, if necessary.
- [ ] Code generation (`make codegen`) has been run to update object specifications, if necessary.
- [ ] Comments have been added or updated based on the standards set in [CONTRIBUTING.md](https://github.com/rook/rook/blob/master/CONTRIBUTING.md#comments)
- [ ] Add the flag for skipping the CI if this PR does not require a build. See [here](https://github.com/rook/rook/blob/master/INSTALL.md#skip-ci) for more details.

### CODE REVIEW

**Comment by davidvossel** [comment]
/cc @travisn @kshlm @jarrpa @leseb

We can use this logic to source the rook-ceph related manifests when building the ocs-operator's unified CSV.

**Comment by davidvossel** [comment]
It looks like the build is failing because this PR makes the the operator-sdk and yq tools a build requirement for rook-ceph.

**Comment by davidvossel** [comment]
> It looks like the build is failing because this PR makes the the operator-sdk and yq tools a build requirement for rook-ceph.

alrighty, I was able to make CI happy again.

I added operator-sdk and yq as optional build dependences for the rook-ceph image. The CSV file templates will be built and added to the image as long as the host OS is linux or darwin. Otherwise that step is ignored. The linux/darwin limitation is due to the operator-sdk cli tool (which is used to generate part of the CSV) only having linux/darwin variants available right now.

**Comment by davidvossel** [comment]
@leseb Thanks for the review, I addressed your feedback.

**Comment by davidvossel** [comment]
> Please squash "Address minor review feedback comments" into whatever is relevant and we should be good to go. Thanks.

done

File: cluster/olm/ceph/generate-rook-csv.sh [hunk 1]
```diff
@@ -75,10 +78,11 @@ fi
...
 #############
-YQ_CMD_DELETE=(yq delete -i)
-YQ_CMD_MERGE=(yq merge --inplace --overwrite --append)
-YQ_CMD_WRITE=(yq write --inplace)
-OP_SDK_CMD=(operator-sdk olm-catalog gen-csv --csv-version)
+YQ_CMD_DELETE=($yq delete -i)
+YQ_CMD_MERGE_OVERWRITE=($yq merge --inplace --overwrite --append)
+YQ_CMD_MERGE=($yq merge --inplace --append)
+YQ_CMD_WRITE=($yq write --inplace)
+OP_SDK_CMD=($operator_sdk olm-catalog gen-csv --csv-version)
```
Comment (travisn) [inline cluster/olm/ceph/generate-rook-csv.sh]: should this command be `operator-sdk` (hyphen instead of an underscore)?

**Comment by davidvossel** [comment]
> The OCS operator won't ever need the templates from the upstream image, right? Shouldn't the downstream build process add these to the downstream image that the OCS operator can pick up?

For now, the templates are relevant for both upstream and downstream. All the ocs-operator integration work is being developed and tested against upstream at the moment.

> What is the overall flow for generating CSVs? Do we have documentation about it for future reference? Does the generate-rook-csh.sh need to be run manually by the dev to generate the CSVs and they are checked in?

No change is required in the dev flows. For the templated csv generation, dev's don't need to be aware it is occurring. It just happens automatically as the image is being built using the example manifests as the source of truth.

For example, if a dev adds a new feature to the ceph operator that involves a new RBAC rule, the only thing that's expected is the example manifests (common.yaml) is updated accordingly. When the image is built, all this csv generation logic automatically picks up the RBAC change and puts it in the templated CSV. The dev didn't need to think about any of that.

> Then the build will run this new script generate-rook-csv-templates.sh to copy the CSVs in to the image? Seems like we should either make them both to be generated by the build, or unify them to a single script to run at the same time.

When and at what cadence you all create the release CSV's is something I didn't want to get involved with. I left that logic and whatever process is behind that logic alone. The csv template generation needs to occur regardless of whether a new CSV release is being cut or not though, which is why I added that flow directly to the build step.

> When building this PR from my mac, i get the following error:
> exec: "x86_64-linux-gnu-gcc": executable file not found in $PATH

I'm not 100% sure what is causing this, but I did fix something related to host platform detection.

If you're still hitting this, can you run `go get gopkg.in/mikefarah/yq.v2` and see what happens when you execute the yq.v2 binary? I'm curious if something is messed up with their darwin build of that tool.

File: cluster/olm/ceph/generate-rook-csv.sh (hunk 1 above)
Comment (davidvossel) [reply cluster/olm/ceph/generate-rook-csv.sh]: this looks right. that's a variable pointing to the operator-sdk binary's path.

File: images/ceph/Makefile [hunk 2]
```diff
@@ -20,24 +20,66 @@ include ../image.mk
...
+ifeq ($(HOST_PLATFORM),linux_amd64)
+OPERATOR_SDK_PLATFORM = x86_64-linux-gnu
+endif
+ifeq ($(PLATFORM),linux_arm64)
+# host os not supported
+endif
+ifeq ($(PLATFORM),darwin_amd64)
+OPERATOR_SDK_PLATFORM = x86_64-apple-darwin
+endif
+ifeq ($(PLATFORM),windows_amd64)
```
[the remark this answers is not in the document]
Comment (davidvossel) [reply images/ceph/Makefile]: fixed. thanks

**Comment by travisn** [comment]
If I run `go get` explicitly as you suggested, it works fine and i am able to run the tool on mac. So something isn't quite right with the line that calls `go get`, but I don't see where the problem is either.

**Comment by davidvossel** [comment]
> If I run go get explicitly as you suggested, it works fine and i am able to run the tool on mac. So something isn't quite right with the line that calls go get, but I don't see where the problem is either.

Without a mac build host, I don't have much more to go on here. 

For now I've limited the csv template generation for linux only. Is that acceptable?

**Comment by travisn** [comment]
> For now I've limited the csv template generation for linux only. Is that acceptable?

Yeah, that seems reasonable. Local dev builds wouldn't rely on this.

**Comment by davidvossel** [comment]
> Yeah, that seems reasonable. Local dev builds wouldn't rely on this.

Is there anything else from my side that this PR needs in order to move forward?

File: images/ceph/Makefile [hunk 3]
```diff
@@ -20,24 +20,60 @@ include ../image.mk
...
-do.build:
+do.build: generate-csv-ceph-templates
 	@echo === docker build $(CEPH_IMAGE)
 	@cp Dockerfile $(TEMP)
 	@cp toolbox.sh $(TEMP)
 	@cp $(OUTPUT_DIR)/bin/linux_$(GOARCH)/rook $(TEMP)
 	@cp $(OUTPUT_DIR)/bin/linux_$(GOARCH)/rookflex $(TEMP)
 	@cp -r ../../cluster/examples/kubernetes/ceph/csi/template $(TEMP)/ceph-csi
 	@cp -r ../../cluster/examples/kubernetes/ceph/monitoring $(TEMP)/ceph-monitoring
+	# Copy temlated csv manifest only if operator-sdk is supported for the host platform.
```
Comment (travisn) [inline images/ceph/Makefile]: I see this line printed with every local build (typo: templated). The intention wasn't to print it, right?
```
=== docker build build-cf4cd2f5/ceph-amd64
# Copy temlated csv manifest only if operator-sdk is supported for the host platform.
sha256:b19a5c6e16acaaa8094886961eebecd45d593e89e3fb67d418b82d3512bd4d36
```

Comment (davidvossel) [reply images/ceph/Makefile]: fixed and updated pr.

**Pull request merged**